feat(rum): propagate sampling decision through tracestate

vigneshshanmugam opened the pull request:
+ fix #845 
+ Adds a new option `propagateTracestate` boolean to optionally add the tracestate header to all outgoing requests from the browser that are captured as transactions and spans . 
+ sent as `sample_rate` and `sr` in both v2 and v3 spec in the APM server. 
+ Updated the config doc and also CORS docs.

apmmachine commented:
### :package: Bundlesize report

Filename | Size(bundled) | Size(gzip) | Diff(gzip)
 --- | --- | --- | ---
elastic-apm-opentracing.umd.min.js | 62.1 KiB | 19.8 KiB |  :warning: 104 Bytes 
elastic-apm-rum.umd.min.js | 56.1 KiB | 18.3 KiB |  :warning: 101 Bytes 

<!--COMMENT_GENERATED_WITH_ID_bundlesize-->

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-rum-js_953.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #953 opened`



* * Start Time: 2021-01-25T23:01:12.061+0000


* Duration: 17 min 45 sec


* Commit: 5454369cc9acb7319272051bbdbd479a5ca0d310



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 3  |
| Passed       | 189  |
| Skipped      | 0 |
| Total        | 192   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![3](https://img.shields.io/badge/3%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Initializing / Test Pupperteer / Matrix - STACK_VERSION = '8.0.0-SNAPSHOT', SCOPE = '@elastic/apm-rum-core' / ApmServer should compress all events when apiVersion is >2 – HeadlessChrome_89_0_4389_(Linux_0_0_0).ApmServer`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
    at <Jasmine>
    at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
    at <Jasmine>
 
 ```
</p></details>
</ul>
    
##### `Initializing / Test Pupperteer / Matrix - STACK_VERSION = '7.0.0', SCOPE = '@elastic/apm-rum-core' / ApmServer should compress all events when apiVersion is >2 – HeadlessChrome_89_0_4389_(Linux_0_0_0).ApmServer`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
    at <Jasmine>
    at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
    at <Jasmine>
 
 ```
</p></details>
</ul>
    
##### `Initializing / Test Pupperteer / Matrix - STACK_VERSION = '7.7.0', SCOPE = '@elastic/apm-rum-core' / ApmServer should compress all events when apiVersion is >2 – HeadlessChrome_89_0_4389_(Linux_0_0_0).ApmServer`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
    at <Jasmine>
    at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
    at <Jasmine>
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->







### Steps errors [![15](https://img.shields.io/badge/15%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-953/detail/PR-953/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  > Show only the first 10 steps failures
  
  
  
  
##### `Run tests: Elastic Stack 7.7.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 4 min 50 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1268/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Run tests: Elastic Stack 7.7.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 1 min 22 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1506/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Run tests: Elastic Stack 7.7.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 1 min 24 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1696/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Archive the artifacts`
<ul>
<li>Took 0 min 1 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1766/log/?start=0">here</a></li>
<li>Description: <code>[2021-01-25T23:17:51.116Z] Archiving artifacts
script returned exit code 1</code></l1>
</ul>
  
  
  
  
##### `General Build Step`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1780/log/?start=0">here</a></li>
<li>Description: <code>[2021-01-25T23:17:53.267Z] Archiving artifacts
script returned exit code 1</code></l1>
</ul>
  
  
  
  
##### `Run tests: Elastic Stack 7.0.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 4 min 49 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1227/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Run tests: Elastic Stack 7.0.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 2 min 31 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1367/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Run tests: Elastic Stack 7.0.0 - @elastic/apm-rum-core - none`
<ul>
<li>Took 2 min 31 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1692/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Archive the artifacts`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1735/log/?start=0">here</a></li>
<li>Description: <code>[2021-01-25T23:17:45.187Z] Archiving artifacts
script returned exit code 1</code></l1>
</ul>
  
  
  
  
##### `General Build Step`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/apm-agent-rum-mbp/pipelines/PR-953/runs/1/steps/1749/log/?start=0">here</a></li>
<li>Description: <code>[2021-01-25T23:17:47.338Z] Archiving artifacts
script returned exit code 1</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-25T23:17:49.705Z] node-puppeteer         |     ✓ should be called after tasks
[2021-01-25T23:17:49.705Z] node-puppeteer         |     ✓ edge case test with double timeout
[2021-01-25T23:17:49.705Z] node-puppeteer         |   ApmServer
[2021-01-25T23:17:49.705Z] node-puppeteer         |     ✓ should not send events when there are no events
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should report http errors when server is down
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should queue items before sending
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should not add any items to queue when not initialized
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should log parse error when response is invalid
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should report http errors for queued events
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should throttle adding events to the queue every minute
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should ignore undefined filtered payload
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should set metadata from config along with defaults
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✓ should ndjson all events
[2021-01-25T23:17:49.706Z] node-puppeteer         |     ✗ should compress all events when apiVersion is >2
[2021-01-25T23:17:49.706Z] node-puppeteer         | 	Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
[2021-01-25T23:17:49.706Z] node-puppeteer         | 	    at <Jasmine>
[2021-01-25T23:17:49.706Z] node-puppeteer         | 	    at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
[2021-01-25T23:17:49.706Z] node-puppeteer         | 	    at <Jasmine>
[2021-01-25T23:17:49.706Z] node-puppeteer         | HeadlessChrome 89.0.4389 (Linux 0.0.0): Executed 15 of 274 (1 FAILED) (0.154 secs / 0.074 secs)
[2021-01-25T23:17:49.707Z] node-puppeteer         | TOTAL: 1 FAILED, 14 SUCCESS
[2021-01-25T23:17:49.707Z] node-puppeteer         | 1) should compress all events when apiVersion is >2
[2021-01-25T23:17:49.707Z] node-puppeteer         |      ApmServer
[2021-01-25T23:17:49.707Z] node-puppeteer         |      Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
[2021-01-25T23:17:49.707Z] node-puppeteer         |     at <Jasmine>
[2021-01-25T23:17:49.707Z] node-puppeteer         |     at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
[2021-01-25T23:17:49.707Z] node-puppeteer         |     at <Jasmine>
[2021-01-25T23:17:49.707Z] node-puppeteer         | HeadlessChrome 89.0.4389 (Linux 0.0.0) failed specs:
[2021-01-25T23:17:49.707Z] node-puppeteer         |   ApmServer > should compress all events when apiVersion is >2
[2021-01-25T23:17:49.707Z] node-puppeteer         | 	Error: Expected $[2] = '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true}}' to equal '{"x":{"id":"transaction-id-0","tid":"trace-id-0","n":"transaction #0","t":"transaction","d":990,"c":null,"m":null,"me":[{"sa":{"xdc":{"v":1},"xds":{"v":990},"xbc":{"v":1}}},{"y":{"t":"app"},"sa":{"ysc":{"v":1},"yss":{"v":980}}},{"y":{"t":"type"},"sa":{"ysc":{"v":1},"yss":{"v":10}}}],"y":[{"id":"span-id-0-1","n":"name","t":"type","s":10,"d":10,"c":null,"su":"subtype"}],"yc":{"sd":1},"sm":true,"sr":0.1}}'.
[2021-01-25T23:17:49.707Z] node-puppeteer         | 	    at <Jasmine>
[2021-01-25T23:17:49.707Z] node-puppeteer         | 	    at UserContext.<anonymous> (webpack:///test/common/apm-server.spec.js:341:1 <- test/common/apm-server.spec.js:8009:9)
[2021-01-25T23:17:49.707Z] node-puppeteer         | 	    at <Jasmine>
[2021-01-25T23:17:49.707Z] node-puppeteer         | Running cleanups: 1
[2021-01-25T23:17:49.707Z] node-puppeteer         | npm ERR! code ELIFECYCLE
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! errno 1
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! @elastic/apm-rum-core@5.8.0 script: `node ../../dev-utils/run-script.js "runSauceTests" "packages/rum-core" "true" "test:unit"`
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! Exit status 1
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! 
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! Failed at the @elastic/apm-rum-core@5.8.0 script script.
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm WARN Local package.json exists, but node_modules missing, did you mean to install?
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! A complete log of this run can be found in:
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR!     /app/.npm/_logs/2021-01-25T23_17_49_066Z-debug.log
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! code ELIFECYCLE
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! errno 1
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! @elastic/apm-rum-core@5.8.0 test:sauce: `npm run script runSauceTests packages/rum-core true test:unit`
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! Exit status 1
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! 
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! Failed at the @elastic/apm-rum-core@5.8.0 test:sauce script.
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm WARN Local package.json exists, but node_modules missing, did you mean to install?
[2021-01-25T23:17:49.708Z] node-puppeteer         | npm ERR! A complete log of this run can be found in:
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR!     /app/.npm/_logs/2021-01-25T23_17_49_085Z-debug.log
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! code ELIFECYCLE
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! errno 1
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! @elastic/apm-rum-core@5.8.0 test: `npm run test:sauce`
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! Exit status 1
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! 
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! Failed at the @elastic/apm-rum-core@5.8.0 test script.
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm WARN Local package.json exists, but node_modules missing, did you mean to install?
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! A complete log of this run can be found in:
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR!     /app/.npm/_logs/2021-01-25T23_17_49_103Z-debug.log
[2021-01-25T23:17:49.709Z] node-puppeteer         | lerna ERR! npm run test exited 1 in '@elastic/apm-rum-core'
[2021-01-25T23:17:49.709Z] node-puppeteer         | npm ERR! code ELIFECYCLE
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! errno 1
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! elastic-apm-rum@0.0.0-monorepo test: `npm run build && lerna run test --scope=$SCOPE --no-prefix  --concurrency=1`
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! Exit status 1
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! 
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! Failed at the elastic-apm-rum@0.0.0-monorepo test script.
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2021-01-25T23:17:49.710Z] node-puppeteer         | 
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR! A complete log of this run can be found in:
[2021-01-25T23:17:49.710Z] node-puppeteer         | npm ERR!     /app/.npm/_logs/2021-01-25T23_17_49_128Z-debug.log
[2021-01-25T23:17:49.710Z] node-puppeteer exited with code 1
[2021-01-25T23:17:49.710Z] Aborting on container exit...
[2021-01-25T23:17:50.179Z] + chmod 755 docker-logs.sh
[2021-01-25T23:17:50.504Z] + ./docker-logs.sh @elastic/apm-rum-core-7_7_0 docker-compose.yml
[2021-01-25T23:17:51.116Z] Archiving artifacts
[2021-01-25T23:17:52.911Z] Post stage
[2021-01-25T23:17:52.941Z] Recording test results
[2021-01-25T23:17:53.267Z] Archiving artifacts
[2021-01-25T23:17:53.554Z] Failed in branch Matrix - STACK_VERSION = '7.7.0', SCOPE = '@elastic/apm-rum-core'
[2021-01-25T23:17:54.188Z] Stage "Stack 8.0.0-SNAPSHOT SauceLabs" skipped due to earlier failure(s)
[2021-01-25T23:17:54.283Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-01-25T23:17:54.380Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-01-25T23:17:54.483Z] Stage "Coverage" skipped due to earlier failure(s)
[2021-01-25T23:17:54.588Z] Stage "Release" skipped due to earlier failure(s)
[2021-01-25T23:17:54.635Z] Stage "Release" skipped due to earlier failure(s)
[2021-01-25T23:17:54.747Z] Stage "Release" skipped due to earlier failure(s)
[2021-01-25T23:17:54.853Z] Stage "Release" skipped due to earlier failure(s)
[2021-01-25T23:17:55.037Z] Stage "Opbeans" skipped due to earlier failure(s)
[2021-01-25T23:17:55.458Z] Running on Jenkins in /var/lib/jenkins/workspace/ent-rum_apm-agent-rum-mbp_PR-953
[2021-01-25T23:17:55.556Z] [INFO] getVaultSecret: Getting secrets
[2021-01-25T23:17:55.922Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-25T23:17:56.856Z] + chmod 755 generate-build-data.sh
[2021-01-25T23:17:56.856Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-953/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-953/runs/1 FAILURE 1004529
[2021-01-25T23:17:57.407Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-953/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-25T23:17:59.890Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-953/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-25T23:18:00.140Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-953/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

jahtalab reviewed (commented):
Thanks @vigneshshanmugam , a few points:
- I think we still need some validation for `tracestate` to avoid sending a large header by mistake or a similar bug (Also adding a test for this)
- The tests are failing at the moment
- I will review the spec in a bit more detail and give this another review.

vigneshshanmugam commented:
@jahtalab I have added minimum validation for the values that we set, i think it should good to catch bugs that can arise due to bad sampling value. Also added and fixed the tests.

jahtalab commented on packages/rum-core/src/common/utils.js:
```diff
@@ -117,6 +117,49 @@ function isDtHeaderValid(header) {
 }
 
+function getTSHeaderValue(span) {
+  if (span.sampleRate == null) {
```
Check for sampleRate being a number

jahtalab commented on packages/rum-core/test/performance-monitoring/performance-monitoring.spec.js:
```diff
@@ -265,9 +265,40 @@ describe('PerformanceMonitoring', function() {
+    expect(req.setRequestHeader).toHaveBeenCalledTimes(2)
+    expect(req.setRequestHeader.calls.argsFor(1)).toEqual([
+      'tracestate',
+      traceStateValue
```
It's better to hardcode the expected value here.

vigneshshanmugam replied:
Done

jahtalab reviewed (changes requested):
@vigneshshanmugam , I've added a few comments:

The spec mentions this as well, which is currently missing:
> For non-sampled transactions set the transaction attributes sampled: false and sample_rate: 0, and omit context. No spans should be captured.

Also, the spec mentions storing the `sampe_rate` on spans as well, are we relying on apm-server to do this?
> Agents will record the sampling rate on transactions and spans as sample_rate

vigneshshanmugam commented:
> Also, the spec mentions storing the sampe_rate on spans as well, are we relying on apm-server to do this?

I spoke with Server team yesterday and we would have to do this in the agent side for now as they cant infer in v2 API. I will change the code to propagate the info for spans.

vigneshshanmugam commented on packages/rum-core/src/common/utils.js:
```diff
@@ -117,6 +117,49 @@ function isDtHeaderValid(header) {
+  if (
+    !VALID_VALUE_REGEX.test(sampleRate) ||
+    INVALID_VALUE_REGEX.test(sampleRate)
+  ) {
```
I already checked that out and it's a bit complex than it needs to be. Both should do the same job and this is much smaller. I added tests, let me know if you want to add more tests.

vigneshshanmugam commented on docs/distributed-tracing-guide.asciidoc:
```diff
@@ -29,7 +29,20 @@ var apm = initApm({
+
+By default, the RUM agent does not propagate the `tracestate` HTTP header to the configured origins. However, 
+the user can change that by enabling the <<propagate-tracestate, propagateTracestate>> flag which
+effectively tells the agent to add `tracestate` HTTP header to the configured origins.
```
Done

jahtalab commented:
@vigneshshanmugam , would you please take a look at the failing tests as well?

vigneshshanmugam commented on packages/rum-core/src/performance-monitoring/transaction.js:
```diff
@@ -172,8 +174,10 @@ class Transaction extends SpanBase {
+  resetFields() {
     this.spans = []
+    this.sampleRate = 0
+    this.context = null
```
Good point, let's remove that.

jahtalab reviewed (approved):
LGTM
